tempesta_fw/http_parser.c

```diff
@@ -1812,26 +1812,103 @@ __req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len)
 
 	__FSM_START(parser->_i_st);
 
+	__FSM_STATE(Req_I_WS) {
+		if (IS_WS(c))
+			__FSM_I_MOVE(Req_I_WS);
+		/* Fall through. */
+	}
+
 	__FSM_STATE(Req_I_Accept) {
-		TRY_STR("text/html", Req_I_Accept, Req_I_AcceptHtml);
-		TRY_STR("*/*", Req_I_Accept, Req_I_AcceptHtml);
+		TRY_STR("text", Req_I_Accept, Req_I_TextSlash);
+		TRY_STR_INIT();
+		if (c == '*')
+			__FSM_I_MOVE(Req_I_StarSlash);
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_Type);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_TextSlash) {
+		if (c == '/')
+			__FSM_I_MOVE(Req_I_TextSlashHtml);
+
+		__FSM_I_MOVE(Req_I_Slash);
+	}
+
+	__FSM_STATE(Req_I_TextSlashHtml) {
+		TRY_STR("html", Req_I_TextSlashHtml, Req_I_AcceptHtml);
 		TRY_STR_INIT();
-		__FSM_I_JMP(Req_I_AcceptOther);
+		__FSM_I_JMP(Req_I_Subtype);
+	}
+
+	__FSM_STATE(Req_I_StarSlash) {
+		if (c == '/')
+			__FSM_I_MOVE(Req_I_StarSlashStar);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_StarSlashStar) {
+		if (c == '*')
+			__FSM_I_MOVE(Req_I_AcceptHtml);
+		return CSTR_NEQ;
 	}
 
 	__FSM_STATE(Req_I_AcceptHtml) {
 		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)) {
 			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
 			__FSM_I_JMP(I_EoT);
 		}
-		__FSM_I_MOVE(Req_I_AcceptOther);
+		__FSM_I_JMP(Req_I_Subtype);
+	}
+
+	__FSM_STATE(Req_I_Type) {
+		__FSM_I_MATCH_MOVE(token, Req_I_Type);
+		c = *(p + __fsm_sz);
+		if (c == '/')
+			__FSM_I_MOVE_n(Req_I_Slash, __fsm_sz + 1);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Slash) {
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_Subtype);
+		if (c == '*')
+			__FSM_I_MOVE(I_EoT);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Subtype) {
+		__FSM_I_MATCH_MOVE(token, Req_I_Subtype);
+		c = *(p + __fsm_sz);
+		if (IS_WS(c) || c == ',' || c == ';')
+			__FSM_I_MOVE_n(I_EoT, __fsm_sz);
+		if (IS_CRLF(c)) {
+			return __data_off(p + __fsm_sz);
+		}
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_QValue) {
+		if (('0' <= c && c <= '9') || c == '.')
+			__FSM_I_MOVE(Req_I_QValue);
+		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c))
+			__FSM_I_JMP(I_EoT);
```

Comment:
JMP looks incorrect since it allows `text/plain; q=0.5 *` (without required `,`) since on the space you jump to `IoT`, moves to `IoT` again on the space, and jump on `*` to `Req_I_Accept`. The same problem we had in `Req_I_AcceptHtml` state. Please also fix and add the two test cases for both the states that we do not allow such values.

Reply:
The comment is not addressed. Jump is not accurate: `,` means that a new type/subtype is expected, `;` - a new parameter is expected.

Reply:
We do not eat the current character now, when we go to `I_EoT` and analyze it in `I_EoT`

Reply:
The conditions `IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)` just doubles conditions in `I_EoT` state, so an unconditional jump should be here - there is no sense to check the same `c` against the same condition twice.